components/renku_data_services/users/kc_api.py

```diff
@@ -76,15 +76,17 @@ def get_users(self) -> Iterable[Dict[str, Any]]:
             "max": self.result_per_request_limit + 1,
         }
         first = 0
-        res = self._http_client.get(url, params={**query_args, "first": first})
-        output = res.json()
-        if not isinstance(output, list):
-            raise ValueError("Received unexpected response from Keycloak for users")
-        output = cast(List[Dict[str, Any]], output)
-        yield from output
-        if len(output) < self.result_per_request_limit + 1:
-            return
-        first += self.result_per_request_limit
+        while True:
+            res = self._http_client.get(url, params={**query_args, "first": first})
+            output = res.json()
+            if not isinstance(output, list):
+                raise ValueError("Received unexpected response from Keycloak for users")
+            output = cast(List[Dict[str, Any]], output)
+            yield from output[:-1]
```

Comment:
wouldn't this ignore/not return the last entry in the list if the page isn't the last? it's not really clear from the code why this is necessary

Reply:
Also we now have three pieces of code that are exactly the same (the while loop), just with different query args. Could be a function that is called with `yield from`

Reply:
So I request one more than what I use on every page. The extra result is used to see whether I need to request the following page or I have reached the end. Because if I receive less than max_requests + 1 then I know there are no more pages available.

And I do yield the last element further below when I know that I am on the last page.

When I am iterating over pages the first element on the current page was the "+1" from the previous page.